File: act col/04/Exercise01.java
// =================================================================
//
// File: Exercise01.java
// Authors: 
//      Samuel Octavio González Azpeitia - A01704696
//      Martín Adrian Noboa Monar - A01704052
// Description: This file contains the code to count the number of
//				even numbers within an array using Java's Fork-Join 
//				technology.
//
// Copyright (c) 2023 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.util.concurrent.RecursiveTask;
import java.util.concurrent.ForkJoinPool;

public class Exercise01 extends RecursiveTask<Integer> {
	private static final int SIZE = 100_000_000;
	private static final int MIN = 10_000;
	private int array[], start, end;

	public Exercise01(int start, int end, int array[]) {
		this.array = array;
		this.start = start;
		this.end = end;
		// place your code here
	}

	// place your code here

    public Integer computeDirectly(){
        int result = 0;
		for (int i = start; i < end; i++){
			if (array[i] % 2 == 0){
				result ++;
			}
		}
		return result;
        
    }

	@Override
    protected Integer compute(){
      if ((end - start) <= MIN) {
			return computeDirectly();
		} else {
			int mid = start + ((end - start) / 2);
			Exercise01 lowerMid = new Exercise01(start, mid, array);
			lowerMid.fork();
			Exercise01 upperMid = new Exercise01(mid, end, array);
			return (upperMid.compute() + lowerMid.join());
		}
    }



	public static void main(String args[]) {
		int array[] = new int[SIZE];
		int result = 0;
		long startTime, stopTime;
		double elapsedTime = 0;
		ForkJoinPool pool;

		Utils.fillArray(array);
		Utils.displayArray("array", array);

		elapsedTime = 0;
		System.out.printf("Starting...\n");
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			pool = new ForkJoinPool(Utils.MAXTHREADS);
			result = pool.invoke(new Exercise01(0, array.length, array));

			stopTime = System.currentTimeMillis();

			elapsedTime += (stopTime - startTime);
		}
		System.out.printf("result = %d\n", result);
		System.out.printf("avg time = %.5f ms\n", (elapsedTime / Utils.N));
	}
}

File: act col/05/Exercise01.cpp
// =================================================================
//
// File: exercise01.cpp
// Author(s):
//      Martin Adrian Noboa Monar - A01704052
//      Samuel Octavio González Azpeitia - A01704696
// Description: This file contains the code to count the number of
//				even numbers within an array using the OpenMP 
//				technology. To compile:
//				g++ -o app -fopenmp exercise01.cpp
//
//			Sin Threads		Con Threads		Speed Up
//			636.083  ms		59.044   ms		10.77
// =================================================================

#include <iostream>
#include <iomanip>
#include <chrono>
#include <omp.h>
#include "utils.h"

using namespace std;
using namespace std::chrono;

// array size
#define SIZE 100000000

// implement your code
int calculate (int *array, int size){
	int result = 0;

	#pragma omp parallel for shared(array,size) reduction(+ : result)
	for (int i = 0; i < size; i++) {
        if (array[i] % 2 == 0) {
            result++;
        }
    }
	return result;
}

int main(int argc, char* argv[]) {
	int *array, result;
	// These variables are used to keep track of the execution time.
	high_resolution_clock::time_point start, end;
	double timeElapsed;

	array = new int[SIZE];
	fill_array(array, SIZE);
	display_array("array", array);
	

	cout << "Starting...\n";
	timeElapsed = 0;
	for (int j = 0; j < N; j++) {
		start = high_resolution_clock::now();

		// call the implemented function
		result = calculate(array, SIZE);

		end = high_resolution_clock::now();
		timeElapsed += duration<double, std::milli>(end - start).count();
	}
	cout << "result = " << result << "\n";
	cout << "avg time = " << fixed << setprecision(3) 
		 << (timeElapsed / N) <<  " ms\n";

	delete [] array;
	return 0;
}